### PR TITLE
added empty step validation

### PR DESCRIPTION
**Pull request by rahulii**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds more validation to cl2 config
**Which issue(s) this PR fixes**:

Part of #1746 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: clusterloader2/api/validation.go [hunk 1]
```diff
@@ -55,9 +55,17 @@ func (v *ConfigValidator) Validate() *errors.ErrorList {
 		allErrs = append(allErrs, v.validateTuningSet(c.TuningSets[i], field.NewPath("tuningSets").Index(i))...)
 	}
 
+	// validate steps
+	isStepEmpty := len(c.Steps) == 0
+	if isStepEmpty  {
```
Comment (marseel) [inline clusterloader2/api/validation.go:60]: Could you just check if number of steps is 0 in this if?

File: clusterloader2/api/validation.go [hunk 2]
```diff
@@ -55,9 +55,17 @@ func (v *ConfigValidator) Validate() *errors.ErrorList {
...
+		allErrs = append(allErrs, field.Invalid(field.NewPath("steps"),0,"cannot be empty"))
+
```
Comment (marseel) [inline clusterloader2/api/validation.go:62]: nit: remove empty line

File: clusterloader2/api/validation.go [hunk 3]
```diff
@@ -55,9 +55,17 @@ func (v *ConfigValidator) Validate() *errors.ErrorList {
...
+	}
+	else {
```
Comment (marseel) [inline clusterloader2/api/validation.go:64]: You don't need this else. If there are no steps, this "for" loop below will do nothing.

**Comment by rahulii** [comment]
@marseel , I have implemented the changes requested!

**Comment by rahulii** [comment]
@marseel I'm not sure why tests are failing!

**Comment by marseel** [comment]
```
api/validation.go:59: File is not `gofmt`-ed with `-s` (gofmt)
```

You need to do
```
```

**Comment by rahulii** [comment]
@marseel ,  so while making chnages to unit test, i'm stuck : 
```
{
			name: "valid config",
			input: Config{
				AutomanagedNamespaces: 10,
				Namespace: NamespaceConfig{
					Number: 1,
				},
				Steps: []*Step{
					Step{
						Phases: []*Phase{
							{
								ReplicasPerNamespace: 10,
							},
						},
						Measurements: []*Measurement{
							{
								Method: "test1",
							},
						},
					},
				},
			},
			expected: nil,
		},
```
it says `cannot use Step{...} (type Step) as type *Step in slice literal` . 
Any hint would be helpful!

**Comment by lyzs90** [comment]
> it says `cannot use Step{...} (type Step) as type *Step in slice literal` .
> Any hint would be helpful!

The correct type is actually an address i.e. &Step{} but the linter will just tell you to drop it:
```golang
Steps: []*Step{
	{
		Phases: []*Phase{
			{
				ReplicasPerNamespace: 10,
			},
		},
		Measurements: []*Measurement{
			{
				Method: "test1",
			},
		},
	},
},
```

**Comment by rahulii** [comment]
@lyzs90 @marseel , in  validate.go we are checking 
```
if !isPhasesEmpty && !isMeasurementsEmpty {
		allErrs = append(allErrs, field.Invalid(fldPath.Child("measurements"), s.Measurements, "can't be non-empty with non-empty phases"))
	}
```
So,I am adding only Phases in Steps, as it will suffice and check for non-empty steps!

**Comment by rahulii** [comment]
Should I also add new unit test for just explicitly testing empty steps ?

**Comment by marseel** [comment]
> Shem Leong Marcel Zięba , in validate.go we are checking
> 
> ```
> if !isPhasesEmpty && !isMeasurementsEmpty {
> 		allErrs = append(allErrs, field.Invalid(fldPath.Child("measurements"), s.Measurements, "can't be non-empty with non-empty phases"))
> 	}
> ```
> 
> So,I am adding only Phases in Steps, as it will suffice and check for non-empty steps!

Sounds good.

> Should I also add new unit test for just explicitly testing empty steps ?

Yes, please and thanks for contribution :)

**Comment by marseel** [comment]
/lgtm
But please squash your commits ;)

**Comment by jkaniuk** [comment]
Thanks!

/approve
